runtime/libia2/threads.c

```diff
@@ -66,12 +65,35 @@ void *ia2_thread_begin(void *arg) {
       : "=a"(result)
       : [fn] "r"(fn), [data] "r"(data), [new_sp_addr] "r"(new_sp_addr)
       : "rdi");
-  /* clang-format on */
-  return result;
 #elif defined(__aarch64__)
-#warning "libia2 does not implement ia2_thread_begin yet"
-  __builtin_trap();
+  __asm__ volatile(
+        // Load argument
+        "mov x0, %[data]\n"
+        // Copy stack pointer to x10
+        "mov x10, sp\n"
+        // Load the stack pointer for this compartment's stack
+        "ldr x11, [%[new_sp_addr]]\n"
+        "mov sp, x11\n"
+        // Push the old stack pointer
+        "str x10, [sp, #-8]!\n"
+        "str x10, [sp, #-8]!\n"
+        // Call fn(data)
+        "blr %[fn]\n"
+        // Pop the old stack pointer
+        "ldr x10, [sp], #-8\n"
```

Comment:
This doesn't look right. Shouldn't it be `#8`? e.g. a aarch64-linux-gnu-gcc invocation generates
```
000000000000072c <_fini>:
 72c:	d503201f 	nop
 730:	a9bf7bfd 	stp	x29, x30, [sp, #-16]!
 734:	910003fd 	mov	x29, sp
 738:	a8c17bfd 	ldp	x29, x30, [sp], #16
 73c:	d65f03c0 	ret
```

Reply:
Good catch, but the problem isn't actually the value of the constant--I got confused with AArch64 pre/post-offset syntax and thought this was loading `*(sp-8)` (pseudo-C). But it actually loads `*sp` then decreases sp by 8. Because the next instruction movs into `sp`, any post-offset here does nothing.

So what we should write is `ldr x10, [sp]`, to load the result of latter of the two `str`s, or `ldr x10, [sp, #8]` to load the result of the former one.

I think maybe the best option would be to `stp x10, xzr, [sp, #-16]!` to push the stack pointer and a zero, then `ldr x10, [sp]` to load only the former of the two.

Reply:
I've pushed a commit doing the latter.